src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs

```diff
@@ -20,75 +23,170 @@ internal static class HPackEncoder
+        {
+            Debug.Assert(values != null);
+            Debug.Assert(values.Length > 0);
+            if (values.Length == 1)
+            {
+                return EncodeHeader(name, values[0], null, null, buffer, out length);
+            }
+
+            // When we have more values, separator must be provided.
+            Debug.Assert(separator != null);
+            return EncodeHeader(name, null, values, separator, buffer, out length);
+        }
+
+        private static bool EncodeHeader(string name, string value, string[] values, string separator, Span<byte> buffer, out int length)
         {
             int i = 0;
             length = 0;
 
-            if (buffer.Length == 0)
+            // We need at least \0 and twice length plus one octet string
+            if (buffer.Length < 5)
             {
                 return false;
             }
 
             buffer[i++] = 0;
 
-            if (i == buffer.Length)
+            if (!EncodeHeaderName(name, buffer.Slice(i), out int nameLength))
             {
                 return false;
             }
+            i += nameLength;
 
-            if (!EncodeString(name, buffer.Slice(i), out int nameLength, lowercase: true))
+            if (!EncodeHeaderValue(value, values, separator, buffer.Slice(i), out int valueLength))
             {
                 return false;
             }
+            i += valueLength;
 
-            i += nameLength;
+            length = i;
+            return true;
+        }
 
-            if (i >= buffer.Length)
+        // Encode header name. It needs to be lower cased and validity was checked when created.
+        private static bool EncodeHeaderName(string s, Span<byte> buffer, out int nameLength)
+        {
+            const int ToLowerMask = 0x20;
+            nameLength = 0;
+
+            if (!IntegerEncoder.Encode(s.Length, 7, buffer, out int currentIndex))
             {
                 return false;
             }
 
-            if (!EncodeString(value, buffer.Slice(i), out int valueLength, lowercase: false))
+            if (currentIndex + s.Length >= buffer.Length)
             {
                 return false;
             }
 
-            i += valueLength;
+            // TODO: use huffman encoding
+            for (int j = 0; j < s.Length; j++)
+            {
+                // TODO Use ASCII ToUpper when #34144 is ready.
+                buffer[currentIndex++] = (byte)(s[j] | (s[j] >= 'A' && s[j] <= 'Z' ? ToLowerMask : 0));
+            }
+            nameLength = currentIndex;
 
-            length = i;
             return true;
         }
 
-        private static bool EncodeString(string s, Span<byte> buffer, out int length, bool lowercase)
+        // Encode fragment or header value without writing out length.
+        private static bool EncodeStringPart(string s, Span<byte> buffer, ref int currentIndex)
```

Comment:
The pattern for these routines is not to pass `ref int currentIndex`, but instead to pass the Span starting at the appropriate place for the next byte to write, and return the length via `out int length`. 

We should be consistent about the pattern. If there's a good reason to change the pattern (e.g. perf) then we should change it everywhere.

BTW, it seems like there's a bug directly below related to this: the length check does not consider `currentIndex`.